c/tskit/tables.c

```diff
@@ -116,13 +115,8 @@ read_table_cols(kastore_t *store, read_table_col_t *read_cols, size_t num_cols)
                 goto out;
             }
         } else if (!(read_cols[j].options & TSK_COL_OPTIONAL)) {
-            ret = TSK_REQUIRED_COL_NOT_FOUND;
+            ret = TSK_ERR_REQUIRED_COL_NOT_FOUND;
             goto out;
-        } else {
```

Comment:
@benjeffery, I changed the semantics here slightly because we were getting inconsistent behaviour when the first column in a list was optional (so, everything else was expected to have 0 rows). Since we're already setting default values for the incoming columns, it seemed best to just use those values as the indicator and not to use the lengths at all. That's why we're now checking to see if metadata schema was set, rather than passing NULL, -1 through to set_metadata_schema.

Reply:
I don't like this method though, it's too tricky. I think #537 will make things a lot simpler.